### ISSUE TITLE
Use this in a plugin/shortcode?

### ISSUE THREAD

**Issue by jorenmartijn** [jarednova/timber#565]
Hi,

Been reading up on this plugin and I like what I see. As with most of these template engines, the documentation seems to focus mainly on usage within themes (I might have missed something).
I'm creating a plugin which I'd like to make a little more maintainable. It uses the Facebook Graph API and fetches events from a Facebook page and makes them available in a shortcode: ``[events]``.
Using HTML in a PHP string is never going to look pretty, so I'd like to use Twig/Timber to separate code from the view in this shortcode. Is that possible using this plugin?
This is my current code for the shortcode:
```php
function fevents($atts ){
	global $fbevents; // This is the Facebook Graph Object
	$output = get_transient('cstfbevents'); // Transient cache get
	$output = "<div id=\"events\">";
	foreach($fbevents->getProperty('data')->asArray() as $single){ 
		$output .= "<div class=\"event\">";
		$basetime = strtotime($single->start_time);
		$carbontime = Carbon::createFromTimeStamp($basetime, 'Europe/Amsterdam');
		//$output .= "<img src=\"".$single->photos->data[0]->icon."\"/>";
		//if(isset($single->cover->source)){ 	$output .= "<header class=\"eventhead\" style=\"background-image:url('{$single->cover->source}');\"></header>";
		//$output .= "<br/>".$single->cover->offset_x."<br/>"; $output .= $single->cover->offset_y;}
		$output .= "<h4>".$single->name."</h4>";
		if(isset($single->description)){ 	$output .= "<aside>".$single->description."</aside>"; }
		//$output .= "<aside>".$single->start_time."</aside>";
		$output .= "<time>{$carbontime}</time>";
		$output .= "</div>";
	}
	$output .= "</div>";
	set_transient('cstfbevents', $output, HOUR_IN_SECONDS); // Set transient cache

	return $output;
}
add_shortcode( 'events', 'fevents' );

```
Thanks in advance!

**Comment by asecondwill** [comment]
It's really simple to use Timber in plugins.  Here is a shortcode plugin i made earlier this week:

function events_feed(){
  $context = Timber::get_context();
  $events = get_posts(array(
          		'post_type'		=> 'event',
          		'posts_per_page'=> 3,
          		'meta_key'		=> 'event_date',
          		'orderby'		=> 'meta_value_num',
          		'order'			=> 'ASC'
          	));

	foreach ($events as $event){
		$context['events'][] = new TimberPost($event->ID);
	}
	Timber::render('_events.twig', $context);
}
add_shortcode('events_feed', 'events_feed');

You can see a more comprehensive plugin here:

https://github.com/asecondwill/Flexible-Content-Widget

I've just realised they both output, rather than return the html.  not sure that is ideal.

**Comment by jorenmartijn** [comment]
Thanks! And echoing is something I want to avoid, so I may have to look into that. Perhaps there's a way to render the output inside a variable? Then you can output it. I think Blade allows you to do that. By the way, transient caching (a wonderful set of functions I only discovered two days ago in WP) should work fine, right? :)

**Comment by anugupta** [comment]
You can use `Timber::compile()` to save the rendered twig output to a variable

https://github.com/jarednova/timber/wiki/Timber-docs#compile

**Comment by asecondwill** [comment]
thanks for pointing that out @anugupta 

@jorenmartijn  -  you good on this now? might be worth closing to keep the tickets down for Jared.

**Comment by jorenmartijn** [comment]
It seems simple enough. Haven't been able to try it out yet, but I'll do that asap. So yes, it can be closed. Let me have a look at how to close this...

**Comment by jorenmartijn** [comment]
closes #565
